memlimit/memlimit.go

```diff
@@ -29,56 +30,106 @@ var (
 	logger = log.New(io.Discard, "", log.LstdFlags)
 )
 
+type config struct {
+	printf   func(string, ...interface{})
+	ratio    float64
+	provider Provider
+}
+
+// An Option alters the behavior of SetGoMemLimitWithEnv.
+type Option interface {
+	apply(*config)
+}
+
+type optionFunc func(*config)
+
+func (of optionFunc) apply(cfg *config) { of(cfg) }
+
+// WithLogger uses the supplied printf implementation for log output.
+// By default log.Printf.
+func WithLogger(printf func(string, ...interface{})) Option {
+	return optionFunc(func(cfg *config) {
+		cfg.printf = printf
+	})
+}
```

Comment:
The logging patterns will be changed after the Go 1.21 release to use the `log/slog` package. (Go 1.21 is expected to be released in August.)

What about separating the logger part to another PR and re-visiting after the Go 1.21 release?

Reply:
@KimMachineGun, now there is no way to manage logging outside the library and by default the `log` package is used. In the new function `SetGoMemLimitWithOptions`, I propose to give the user the opportunity to use
your logger (e.g. `zap`)

Reply:
I think the `printf` style logger is not as customizable as [slog.Logger](https://pkg.go.dev/golang.org/x/exp/slog#Logger). slog.Logger can even handle log level and logging style (in a structured way such as json, key-value style, etc) as well.

So I want to reserve the name `WithLogger` for `slog.Logger` and avoid adding new API (even in another name) likely to be deprecated soon (after Go 1.21).

Would `slog.Logger` not be sufficient to qualify your needs?